### PR TITLE
Fix #1873: Fix OnBoardingFragmentTest for robolectric

### LINKED ISSUE #1873: Fix OnboardingFragmentTest for Robolectric
Currently, the tests in OnboardingFragmentTest are passing on Espresso but they should also pass on Robolectric. Fix this file to make sure that the tests pass on Robolectric.

You can run robolectric tests by following these steps: https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#running-roboelectric-tests

Fixing test cases will need to introduce TextApplicationComponent. This can be seen [here](https://github.com/oppia/oppia-android/pull/1680/commits/4729d551b86282ce617e1d53f2762416758785a1).

### PR DESCRIPTION
**Pull request by prayutsu**
## Explanation

Fixes #1873 
Failing test cases on robolectric are fixed. 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Robolectric - 
Screenshot from 2020-10-13 14-19-05

Espresso - 
Screenshot from 2020-10-13 15-04-54

### CODE REVIEW

**Comment by prayutsu** [comment]
@MohamedMedhat1998 I tried fixing the test cases by introducing a scrollView in the layouts but it still doesn't work. Can you please take a look at that?

**Comment by prayutsu** [comment]
P.S - TO-DO will get removed in the next commit.

**Comment by prayutsu** [comment]
@MohamedMedhat1998 @rt4914 
I see one solution here. 
Rather checking the title of the `slide_title_text_view` we can check the visibility of `skip_text_view` after clicking on `skip_text_view`
If it isn't visible, we can confirm that we are on the last slide.
Does that seems like a solution ?

**Comment by rt4914** [comment]
> @MohamedMedhat1998 @rt4914
> I see one solution here.
> Rather checking the title of the `slide_title_text_view` we can check the visibility of `skip_text_view` after clicking on `skip_text_view`
> If it isn't visible, we can confirm that we are on the last slide.
> Does that seems like a solution ?

@prayutsu That can work too.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt [hunk 1]
```diff
@@ -531,16 +508,14 @@ class OnboardingFragmentTest {
   }
 
   @Test
-  // TODO(#973): Fix OnboardingFragmentTest
-  @Ignore
   fun testOnboardingFragment_clickOnSkip_changeOrientation_titleIsCorrect() {
     launch(OnboardingActivity::class.java).use {
```
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt:562]: @MohamedMedhat1998 @rt4914 Now only this test is failing

**Comment by prayutsu** [comment]
@rt4914 @MohamedMedhat1998 Implementation in [this](https://github.com/oppia/oppia-android/pull/1960#issuecomment-706274778) comment is working properly so it proves that clicking on `skip_text_view` is working on robolectric, but it is not recognizing the displayed `slide_title_text` properly.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt:562]: @prayutsu Can you mention the error?

Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt:562]: @rt4914 actually it is recognising the text of first onboarding slide i.e - `Welcome to Oppia!` 

I have highlighted the text being recognised (**)


```
androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'with string from resource id: <2131755271>' doesn't match the selected view.
Expected: with string from resource id: <2131755271>[onboarding_slide_3_title] ***value: Have fun!***
     Got: "AppCompatTextView{id=2131296884, res-name=slide_title_text_view, visibility=VISIBLE, width=792, height=41, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@12174f72, tag=null, root-is-layout-requested=false, has-input-connection=false, x=84.0, y=888.0,


 **text=Welcome to Oppia!, **

input-type=0, ime-target=false, has-links=false}"
```

Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt:562]: > @rt4914 actually it is recognising the text of first onboarding slide i.e - `Welcome to Oppia!`
> 
> I have highlighted the text being recognised (**)
> 
> ```
> androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'with string from resource id: <2131755271>' doesn't match the selected view.
> Expected: with string from resource id: <2131755271>[onboarding_slide_3_title] ***value: Have fun!***
>      Got: "AppCompatTextView{id=2131296884, res-name=slide_title_text_view, visibility=VISIBLE, width=792, height=41, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@12174f72, tag=null, root-is-layout-requested=false, has-input-connection=false, x=84.0, y=888.0,
> 
> 
>  **text=Welcome to Oppia!, **
> 
> input-type=0, ime-target=false, has-links=false}"
> ```


The text of the third slide (`Have fun!`) should be recognized and not that of the first slide (`Welcome to Oppia!`)

**Comment by anandwana001** [comment]
@prayutsu  
1. Please add the screenshot showing test cases are passing on both robolectric and espresso in the PR description
2. As per your test case update, you did the other way round. 
Firstly, we were checking if the textView is visible with that particular text in the 3rd slide or not. But, as per your changes, you are checking that the textView from the first slide should not be visible. 
But, In my point of view, this way we cannot verify that we are on 3rd slide or not after click skip button or whatever action we want to get there, it just first slide is not visible. 
You might need to use some delay as there is a difference in the way how espresso runs the test cases and how robolectric does. 

Check our wiki page [here](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric 
) regarding testing.

**Comment by prayutsu** [comment]
@rt4914 @anandwana001 Till now, I have tried fixing the tests by using `testCoroutineDispatchers`, introducing a scrollView in the OnBoardingFragment layout, using `ViewMatchers.Visibility.VISIBLE` instead of `isDispalyed()`.
I think the last mentioned approach isn't correct because all the slides are visible, none of them possess `INVISBLE` or `GONE` visibility.
Any other approach that you think should work?

**Comment by anandwana001** [comment]
`isDispalyed()` also consider the view which is off-screen that's why it's giving multiple views error.

**Comment by anandwana001** [comment]
@prayutsu @rt4914 I tried out a few other approaches [here](https://gist.github.com/anandwana001/726b4b621035872a0322efb2d061ca18), but I need to investigate more on looper and time delay to fix the issue. By the time, if you got anything let me know, might help.

**Comment by BenHenning** [comment]
Per discussion with @anandwana001 during the team meeting: suggest using runUntilIdle if the observed issue is an animation flake.

**Comment by prayutsu** [comment]
> Per discussion with @anandwana001 during the team meeting: suggest using runUntilIdle if the observed issue is an animation flake.

@BenHenning Are you referring to `advancedUntilIdle()`, if yes I have already tried and it's not working

**Comment by BenHenning** [comment]
I did mean advanceUntilIdle.

@anandwana001 could you spend a bit today trying to repro the failing test on both Espresso & Robolectric? I still don't have a good sense from reading this PR why the failure is happening, and we need that information before we can come up with a reasonable solution.

**Comment by anandwana001** [comment]
> I did mean advanceUntilIdle.
> 
> @anandwana001 could you spend a bit today trying to repro the failing test on both Espresso & Robolectric? I still don't have a good sense from reading this PR why the failure is happening, and we need that information before we can come up with a reasonable solution.

On it. Few points by the time:
1. I found one test `testOnboardingFragment_clickOnSkip_changeOrientation_titleIsCorrect` passing when running on legacy mode. 
2. All the test which are failing, does have some issue with skip part animation.
3. Other two failing test cases where we have two operations, does have some timing issue, the first can be achieved with the TestCoroutineDispatchers, but not the second one not. 
```kotlin
onView(withId(R.id.onboarding_slide_view_pager)).perform(scrollToPage(2))
onView(withId(R.id.skip_text_view)).perform(click())
```

Did some assertion check on `skip` text weather the string is visible after clicking on it or not, it's not, and that's correct. I do think we need some better understanding of how views are visible with `ViewPager` as here all views are in visible state as all fragment gets drawn already and `skip` text is something which is on the activity not fragment. 

on the contrary to the above point, after clicking on `skip`, the view is not synced(multiple fragments) and when doing the next check leads to failure.

**Comment by BenHenning** [comment]
Hmm, when running it locally the way that this layout is set up actually just confuses me. It seems like we're using ambiguous IDs to correspond to the text view that's being matched (e.g. there are two slide_title_text_view TextViews, one that's shared for slides 0-2 and one that's used for slide 3). That might be complicating the matching here.

**Comment by BenHenning** [comment]
Interestingly, this works fine if we use scrollToPage(). I think the internal animation triggered when switching pages is breaking this (which would explain why LEGACY mode works here), but I haven't yet figured out why advanceUntilIdle() is insufficient.

**Comment by BenHenning** [comment]
Aha! Confirmed it is scrolling: I disabled smooth scrolling and it passes, e.g.:

```kotlin
  override fun clickOnSkip() {
    binding.onboardingSlideViewPager.setCurrentItem(TOTAL_NUMBER_OF_SLIDES - 1, false)
  }
```

(from OnboardingFragmentPresenter). Now the question is figuring out why. :)

**Comment by BenHenning** [comment]
So I'm getting stumped. My best guess is that smooth scrolling works by taking animation steps every draw call (as driven by the Choreographer), but this doesn't seem to work correctly when using a paused looper. I've run into other issues in the past because of this behavior, and I haven't yet found a solution.

I see a few possible solutions (in order from best to worst):
1. Migrating to ViewPager2 (https://developer.android.com/training/animation/vp2-migration). From some findings I uncovered, ViewPager2 might actually work better with a PAUSED LooperMode, so it might just fix our issue. It also uses RecyclerView internally, and we rely on RecyclerView animations elsewhere, so there's a decent change it will "just work".
2. Mark the test as Espresso only using the ``RunOn`` annotation. This obviously isn't ideal, but it lets us at least get test coverage. We should file an issue to track fixing this test on Robolectric.
3. Enable LEGACY LooperMode just for that test. This is actually less ideal than 2 because it may result in race conditions that trigger flakiness, and LEGACY may be removed in future versions of Robolectric. We also should have an issue filed to remove the legacy part in the future.

I strongly suggest looking into (1) first @prayutsu to see if it fixes the issue. This may lead to UI changes so we be careful to review that for both portrait & landscape on both tablet & phone device sizes to make sure there aren't any regressions.

**Comment by anandwana001** [comment]
@prayutsu 
Few points for migration:
1. `RecyclerView.Adapter` replaces `PagerAdapter`
2. `registerOnPageChangeCallback` replaces `addPageChangeListener`

**Comment by prayutsu** [comment]
@anandwana001 I am stuck on how to modify the new methods. Can you please help here?

**Comment by prayutsu** [comment]
@anandwana001 Please confirm if the PR is progressing in the right direction

**Comment by rt4914** [comment]
Un-assigning for now a this PR is not yet finished.

**Comment by anandwana001** [comment]
> @anandwana001 Please confirm if the PR is progressing in the right direction

yes, i guess, as per our discussion in the meeting have you tried implementing the ViewPager2 in a sample project and try to apply here as well. 

@prayutsu as per my earlier suggestion from earlier PR's, always try to fix all the checks here before any review, it's quite difficult to suggest anything with the failing checks. 

for ktlint, I suggest taking a look at our wiki and set up at your local to prevent any ktlint issues.

**Comment by prayutsu** [comment]
@anandwana001 Thanks, I'll do the same in a couple of days

**Comment by anandwana001** [comment]
@prayutsu I had created a sample repo on implementation of ViewPager2 - https://github.com/anandwana001/sample-viewpager2

**Comment by anandwana001** [comment]
@prayutsu I tried the migration, but need few nit fixes, adding gist here, this might help you, you can carry forward the work. Before you start, we can connect over the meeting. 

https://gist.github.com/anandwana001/8885f37c452fe4fc1a035b14d5fbded4

**Comment by prayutsu** [comment]
@anandwana001 I have my lab today, so I would be available after 6 PM, also as this is a festive week, we both may not manage time for the meeting, but still, we can discuss our preferred time slots on gitter

**Comment by prayutsu** [comment]
@anandwana001 Tried really hard to find a replacement of `scrollToPage()` for ViewPager2, documentation also contains information only about `ViewPagerActions`, I was able to find some examples but they didn't contain something useful but I am still adding the link what I found - https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/viewpager2/integration-tests/testapp/src/androidTest/java/androidx/viewpager2/integration/testapp/test/FakeDragTest.kt

**Comment by prayutsu** [comment]
@anandwana001 After long research, I found something that may work, can you please take a look at it and tell if we can adopt this?
https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/viewpager2/integration-tests/testapp/src/androidTest/java/androidx/viewpager2/integration/testapp/test/PreviewPagesTest.kt

**Comment by anandwana001** [comment]
@prayutsu could you try and implement. Also, I think this is the test for the ViewPager2 core implementation.

**Comment by prayutsu** [comment]
@anandwana001 I tried this, but it's not working, there's no method named `swipeNext()`. Can you please help in figuring out a way to use `srollToPage()` with ViewPager2?

**Comment by anandwana001** [comment]
> @anandwana001 I tried this, but it's not working, there's no method named `swipeNext()`. Can you please help in figuring out a way to use `srollToPage()` with ViewPager2?

@prayutsu ok, I will check this. Let's do one thing, could you fix all these checks once and then I will pull this branch and work forward, it will be a great help. Also, as per our discussion those failing test cases mark them ignore so it will not fail on checks here.

**Comment by prayutsu** [comment]
@anandwana001 Sure, I will do that

**Comment by prayutsu** [comment]
@anandwana001 One non-app module test is failing but I think it will pass if checks are run again. Also, I am not able to understand why the Build with Bazel check is failing, do we need to add something in the `Build.bazel` file?

**Comment by prayutsu** [comment]
I have updated my branch with the latest develop, maybe now all the checks would pass

**Comment by prayutsu** [comment]
@anandwana001 Can you help me figure out why the last check is failing?

**Comment by anandwana001** [comment]
@prayutsu using viewPager2 without dependency and bazel is unable to build with all the required dependency. 
Did you check my pr here - https://github.com/oppia/oppia-android/pull/2104/files ?

**Comment by prayutsu** [comment]
@anandwana001 Ohh, I thought this is merged now, thanks for reminding

**Comment by prayutsu** [comment]
@anandwana001 PTAL

**Comment by prayutsu** [comment]
@anandwana001 I'll be working further on this PR

**Comment by prayutsu** [comment]
Closing this PR as @anandwana001 will take up the issue because robolectric stuff needs to be wrapped quickly

**Comment by prayutsu** [comment]
@rt4914 @BenHenning  We decided to migrate to ViewPager2 because the test cases related to the `skip` button were not working, but even after migrating to `ViewPager2` the test cases are still failing

**Comment by anandwana001** [comment]
> @rt4914 @BenHenning We decided to migrate to ViewPager2 because the test cases related to the `skip` button were not working, but even after migrating to `ViewPager2` the test cases are still failing

I will take a look into this.

**Comment by rt4914** [comment]
> @rt4914 @BenHenning We decided to migrate to ViewPager2 because the test cases related to the `skip` button were not working, but even after migrating to `ViewPager2` the test cases are still failing

@anandwana001 Thanks for helping out. 
Just add one point, we still need ViewPager2 because `ViewPager2` supports RTL by default and `ViewPager` does not support RTL.

**Comment by anandwana001** [comment]
Once the implementation is done with the ViewPager2, then we can go for testing. 
Add me as a reviewer, once the implementation is done.

**Comment by prayutsu** [comment]
@anandwana001 @rt4914  PTAL at this video - 
After migrating to ViewPager2, we have the following regression. I tried to find the error in the new code but couldn't find the culprit. Can you please help with it?
ezgif com-gif-maker

**Comment by prayutsu** [comment]
The bug seems a little weird because after sliding more than 2 times, the page restores its original and valid state but immediately after changing to Landscape orientation, it shows a small image in the top-left corner

**Comment by prayutsu** [comment]
@anandwana001 @rt4914 We are now (after migrating to ViewPager2) facing a regression in OnBoardingSlides. I tried to find the error in the newly added code but after struggling for hours, I couldn't find any errors. I didn't even find anything on the Internet. Can you please help?
ezgif com-gif-maker

**Comment by prayutsu** [comment]
The bug is a bit weird because after sliding more than 2 times, we restore our original and valid configuration

**Comment by anandwana001** [comment]
any update on this?

**Comment by Luffy18346** [comment]
@rt4914 PTAL

**Comment by anandwana001** [comment]
Closing this PR as no further update.

**Pull request closed without merging**